src/ORM/DataList.php

```diff
@@ -1235,39 +1245,46 @@ private function fetchEagerLoadManyMany(
         $childIDField = $manyManyLastComponent['childField'];
-        // $join will either be:
-        // - the join table name for many-many
-        // - the join data class for many-many-through
-        $join = $manyManyLastComponent['join'];
+        $joinTable = $manyManyLastComponent['join'];
+        $extraFields = $manyManyLastComponent['extraFields'];
+        $joinClass = $manyManyLastComponent['joinClass'];
+
+        // Get the join records so we can correctly identify which children belong to which parents
+        $joinRows = DB::query('SELECT * FROM "' . $joinTable . '" WHERE "' . $parentIDField . '" IN (' . implode(',', $parentIDs) . ')');
 
         // many_many_through
-        if (is_a($manyManyLastComponent['relationClass'], ManyManyThroughList::class, true)) {
-            $joinThroughObjs = DataObject::get($join)->filter([$parentIDField => $parentIDs]);
-            $relationItemIDs = [];
-            $joinRows = [];
-            foreach ($joinThroughObjs as $joinThroughObj) {
-                $joinRows[] = [
-                    $parentIDField => $joinThroughObj->$parentIDField,
-                    $childIDField => $joinThroughObj->$childIDField
-                ];
-                $relationItemIDs[] = $joinThroughObj->$childIDField;
-            }
+        if ($joinClass !== null) {
+            $relationList = ManyManyThroughList::create(
+                $relationDataClass,
+                $joinClass,
+                $childIDField,
+                $parentIDField,
+                $extraFields,
+                $relationDataClass,
+                $parentDataClass
+            )->forForeignID($parentIDs);
         // many_many + belongs_many_many
         } else {
-            $joinTableQuery = DB::query('SELECT * FROM "' . $join . '" WHERE "' . $parentIDField . '" IN (' . implode(',', $parentIDs) . ')');
-            $relationItemIDs = $joinTableQuery->column($childIDField);
-            $joinRows = $joinTableQuery;
+            $relationList = ManyManyList::create(
+                $relationDataClass,
+                $joinTable,
+                $childIDField,
+                $parentIDField,
+                $extraFields
+            )->forForeignID($parentIDs);
         }
 
         // Get ALL of the items for this relation up front, for ALL of the parents
+        // Use a real RelationList here so that the extraFields and join record are correctly set for all relations
         // Fetched as a map so we can get the ID for all records up front (instead of in another nested loop)
         // Fetched after that as an array because for some reason that performs better in the loop
         // Note that "Me" is a method on ViewableData that returns $this - i.e. that is the actual DataObject record
-        $relationArray = DataObject::get($relationDataClass)->byIDs($relationItemIDs)->map('ID', 'Me')->toArray();
+        $relationArray = $relationList->map('ID', 'Me')->toArray();
```

Comment:
_all_ of this is just using the `ManyManyList` and `ManyManyThroughList` directly to get the records, instead of crafting our own queries. This means we get the extraFields data and join records set up correctly "for free" (read: without having to add any extra code, but with a ~50ms performance hit)